README.md
# Alex vs Bus: The Race

**Alex vs Bus: The Race** is a platform runner game for Android and PCs with
Java in which Alex is just a man who depends on public transportation and needs
to run in order to catch the bus, or else he will have to wait an eternity for
the next bus to come.

**Note: not officially on Google Play.** Someone has published the game on
Google Play without having contacted us and charges for it. It is not official
and we cannot ensure it is identical to the official build and up to date. At
present, the game is officially distributed only through GitHub, where it is
free of charge.

As it is currently a pre-release, not all planned features are present yet, but
it is already fully playable and has all planned levels.

The graphical and sound style is mostly inspired by the 8-bit era, although the
game is not faithful to all of the technical limitations of the period.

The library used by the game is libGDX (https://libgdx.com) and the source code
repository can be found at https://github.com/M374LX/alexvsbus-java.

![Screenshot 1](./docs/img/screenshot1.png)
![Screenshot 2](./docs/img/screenshot2.png)
![Screenshot 3](./docs/img/screenshot3.png)


## Gameplay

The game can be described as a "platform runner" (not an endless runner). In
each level, the player has 90 seconds to catch the bus while jumping over holes,
banana peels, crates, and other obstacles.


## Controls

The game can be controlled by using a keyboard, gamepad/joystick, or touchscreen
buttons.

When using a keyboard, the controls are:

**Arrow keys/WASD:** Move character; select dialog item

**Ctrl/Spacebar:** Jump; confirm selected dialog item

**Enter (aka Return):** Confirm selected dialog item

**Escape:** Pause; close dialog

**F5:** Toggle fullscreen (PC only)

**F6:** Toggle audio

**F7:** Toggle scanlines

On French (AZERTY) keyboards, W becomes Z and A becomes Q.

On Android, the Back button has the same effect as Escape.

Touchscreen buttons appear by default on Android, while these can be enabled on
PC by using the command line option ``--touch``. Dialog items can also be
selected by touching them.


## Downloading

To avoid the need to build the game from source, it can be downloaded in
pre-built form from https://github.com/M374LX/alexvsbus-java/releases. The .jar
file is for PC, while the .apk file is for Android.


## Running on PC

On PC (including Windows and Linux), the game requires Java (https://java.com).
On Debian, it can be installed by running the command:
```
apt install openjdk-11-jre
```

After installing Java and downloading the .jar file, on Windows, a double-click
on the .jar file should be enough to start the game. To start it from the CLI,
the command is:
```
java -jar alexvsbus-desktop-pre6.jar
```


## Command line options (PC only)

On PC, the game supports some command line options. For example, to use the
option --fullscreen, the full command is:
```
java -jar alexvsbus-desktop-pre6.jar --fullscreen
```

The options are:
```
-h, --help             Show this usage information and exit
-v, --version          Show version and license information and exit
-f, --fullscreen       Run in fullscreen mode
-w, --windowed         Run in windowed mode
--window-scale <scale> Set the window scale (1 to 3)
--audio-on             Enable audio output
--audio-off            Disable audio output
--resizable            Make the window resizable
--scanlines-on         Enable scanlines visual effect
--scanlines-off        Disable scanlines visual effect
--touch                Enable touchscreen controls, which can also be
                       simulated by using the mouse
--touch-buttons-on     Enable left, right, and jump buttons on
                       touchscreen (visible only if --touch is also used)
--touch-buttons-off    Disable left, right, and jump buttons on
                       touchscreen
--vscreen-size <size>  Set the size of the virtual screen (vscreen)

For --vscreen-size, the size can be either "auto" or a width and a height
separated by an "x" (example: 480x270), with the supported values listed
below.

Supported width values:
480
432
416
320
256

Supported height values:
270
256
240
224
192
```


## Building from source

In order to build the game from source, the Java Development Kit (JDK) is
required. Android still requires version 8 of the JDK. It can be downloaded from
https://www.oracle.com/java/technologies/javase/javase-jdk8-downloads.html.

As a result of being a libGDX project, the game uses Gradle as its build system.
The source tree contains the Gradle wrapper, which downloads the appropriate
Gradle version and the dependencies automatically. For this reason, an Internet
connection is required when building the game for the first time.

### PC

The executable .jar file can be built by running the following command on Unix
systems (including Linux):
```
./gradlew desktop:dist
```

On Windows, replace "./gradlew" with "gradlew.bat".

The built .jar file will be at ``desktop/build/libs/desktop-pre6.jar``.

Alternatively, the following command allows running the game without keeping the
build:
```
./gradlew desktop:run
```

### Android

In order to build for Android, the Android SDK is required in addition to the
JDK version 8. It can be downloaded from https://developer.android.com/studio.
The command line tools are enough, without the need for Android Studio.

The required SDK packages, which can be installed by using ``sdkmanager``, are
``build-tools;29.0.3`` and ``platforms;android-29``. The versions here are the
ones that have been tested, but later versions might also work. A few other
packages will be installed automatically as dependencies. The NDK can be
optionally installed to prevent messages like "Unable to strip library due to
missing strip tool" and "Compatible side by side NDK version was not found",
although is does not significantly shrink the .apk file.

The file ``local.properties`` needs to be present and point to the Android SDK
installation directory, like so:
```
sdk.dir=/path/to/android/sdk
```

The command to build the .apk file for Android on Unix systems (including Linux)
is:
```
./gradlew android:assembleDebug
```

On Windows, replace "./gradlew" with "gradlew.bat".

The built .apk file will be at
``android/build/outputs/apk/debug/android-debug.apk``.

At present, only debug builds have been tested.

### Cleaning

To clean the source tree after the project is built, the following command can
be used on Unix systems (including Linux):
```
./gradlew clean
```

On Windows, replace "./gradlew" with "gradlew.bat".


## Project status and plans

While it is quite close to becoming a finished game, there are still plans to
offer the user more settings and improve the documentation, as well as more
testing.

There are also plans to port the game from Java/libGDX to C/SDL. This is why
the GitHub repository contains "java" in its name and also why many Java
features are avoided, including inheritance and generics. The new port will not
be a replacement for this one. Instead, there will be two editions of the game:
the Java/libGDX edition will be focused on Android and the C/SDL version will
be focused on PC while not requiring the JRE.



docs/News.md
## Pre-release: pre6 (2023/04/14)

Sixth pre-release, which includes:

* Improved support for different screen sizes
* Scanlines visual effect
* Immersive mode on Android
* New settings dialog
* No more left side camera locking


## Pre-release: pre5 (2022/09/24)

Fifth pre-release, which includes:

* New "Try again" item on pause dialog
* Fix for the weird "standing on air" visual effects
* Bus movement at level start


## Pre-release: pre4 (2022/04/14)

Fourth pre-release, which includes:

* An ending sequence
* Some visual improvements


## Pre-release: pre3 (2021/12/22)

Third pre-release, which includes:

* New cutscenes
* New "Try again?" dialog
* On the graphics, addition of mirrors and door handles to cars
* Bugfixes


## Pre-release: pre2 (2021/05/18)

Second pre-release, with the new "super" difficulty and a bugfix for game
progress saving on Android.


## Pre-release: pre1 (2021/04/29)

First pre-release.



core/src/org/alexvsbus/LevelLoad.java
/*
 * Alex vs Bus
 * Copyright (C) 2021-2023 M374LX
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 */


package org.alexvsbus;

import static org.alexvsbus.Defs.*;

import com.badlogic.gdx.Gdx;

class LevelLoad {
    PlayCtx ctx;
    LineRead lineRead;

    boolean invalid;

    int xMax;
    int numObjs;
    int numCrateBlocks;
    int numGushes, numGushCracks;
    int numSolids;
    int numHoles, numDeepHoles, numPassageways;
    int numRespawnPoints;
    int numTriggers, numCarTriggers;

    //--------------------------------------------------------------------------

    LevelLoad(PlayCtx ctx) {
        this.ctx = ctx;
        lineRead = new LineRead();
    }

    int load(String filename) {
        boolean noObjects = true;
        int x, y, w, h;
        int i, j;

        invalid = false;

        try {
            if (Gdx.files.internal(filename).length() > 4096) { //4 kB
                return LVLERR_TOO_LARGE;
            }

            lineRead.setData(Gdx.files.internal(filename).readString());
        } catch (Exception e) {
            return LVLERR_CANNOT_OPEN;
        }

        xMax = NONE;
        numObjs = 0;
        numCrateBlocks = 0;
        numGushes = 0;
        numGushCracks = 0;
        numSolids = 0;
        numHoles = 0;
        numDeepHoles = 0;
        numPassageways = 0;
        numRespawnPoints = 0;
        numTriggers = 0;
        numCarTriggers = 0;

        ctx.levelSize = NONE;
        ctx.bgColor = NONE;
        ctx.bgm = NONE;

        x = VSCREEN_MAX_WIDTH_LEVEL_BLOCKS;

        while (!lineRead.endOfData()) {
            String line = lineRead.getLine();
            String tokens[];
            int numTokens;
            int token1, token2, token3, token4;

            if (lineRead.isInvalid()) {
                return LVLERR_INVALID;
            }

            //Skip blank lines
            if (line.length() == 0) {
                continue;
            }

            tokens = line.split(" ");
            numTokens = tokens.length;

            if (numTokens < 2 || numTokens > 5) {
                return LVLERR_INVALID;
            }

            token1 = (numTokens >= 2) ? lineRead.toInt(tokens[1]) : NONE;
            token2 = (numTokens >= 3) ? lineRead.toInt(tokens[2]) : NONE;
            token3 = (numTokens >= 4) ? lineRead.toInt(tokens[3]) : NONE;
            token4 = (numTokens >= 5) ? lineRead.toInt(tokens[4]) : NONE;

            //Invalid integer token
            if (lineRead.isInvalid()) {
                return LVLERR_INVALID;
            }

            if (token1 == NONE) {
                return LVLERR_INVALID;
            }

            if (tokens[0].equals("level-size")) {
                //Error: level size redefinition
                if (ctx.levelSize != NONE) {
                    return LVLERR_INVALID;
                }

                //Error: size out of the allowed range
                if (token1 < 8 || token1 > 32) {
                    return LVLERR_INVALID;
                }

                //Just before the last screen
                xMax = (token1 - 1) * VSCREEN_MAX_WIDTH_LEVEL_BLOCKS;

                ctx.levelSize = token1 * VSCREEN_MAX_WIDTH;

                continue;
            } else if (tokens[0].equals("sky-color")) {
                //Error: sky color redefinition
                if (ctx.bgColor != NONE) {
                    return LVLERR_INVALID;
                }

                switch (token1) {
                    case 1:  ctx.bgColor = SPR_BG_SKY1; break;
                    case 2:  ctx.bgColor = SPR_BG_SKY2; break;
                    case 3:  ctx.bgColor = SPR_BG_SKY3; break;
                    default: return LVLERR_INVALID;
                }

                continue;
            } else if (tokens[0].equals("bgm")) {
                //Error: BGM redefinition
                if (ctx.bgm != NONE) {
                    return LVLERR_INVALID;
                }

                switch (token1) {
                    case 1:  ctx.bgm = BGM1; break;
                    case 2:  ctx.bgm = BGM2; break;
                    case 3:  ctx.bgm = BGM3; break;
                    default: return LVLERR_INVALID;
                }

                continue;
            }

            //Error: adding objects without defining the level size, sky color,
            //and BGM
            if (ctx.levelSize == NONE || ctx.bgColor == NONE || ctx.bgm == NONE) {
                return LVLERR_INVALID;
            }

            //The value of token1 is relative to the previous X position
            x += token1;

            if (tokens[0].equals("banana-peel")) {
                addObj(OBJ_BANANA_PEEL, x, token2, true);
            } else if (tokens[0].equals("car-blue")) {
                addObj(OBJ_PARKED_CAR_BLUE, x, NONE, false);
            } else if (tokens[0].equals("car-silver")) {
                addObj(OBJ_PARKED_CAR_SILVER, x, NONE, false);
            } else if (tokens[0].equals("car-yellow")) {
                addObj(OBJ_PARKED_CAR_YELLOW, x, NONE, false);
            } else if (tokens[0].equals("coin-silver")) {
                addObj(OBJ_COIN_SILVER, x, token2, true);
            } else if (tokens[0].equals("coin-gold")) {
                addObj(OBJ_COIN_GOLD, x, token2, true);
            } else if (tokens[0].equals("crates")) {
                addCrateBlock(x, token2, token3, token4);
            } else if (tokens[0].equals("gush")) {
                addObj(OBJ_GUSH, x, NONE, false);

                if (numGushes >= MAX_GUSHES) {
                    return LVLERR_INVALID;
                }

                ctx.gushes[numGushes].obj = numObjs - 1;
                ctx.gushes[numGushes].y = GUSH_INITIAL_Y;
                ctx.gushes[numGushes].movePattern = gushMovePattern1;
                ctx.gushes[numGushes].movePatternPos = 0;
                ctx.gushes[numGushes].yvel = gushMovePattern1[0];
                ctx.gushes[numGushes].ydest = gushMovePattern1[1];

                numGushes++;
            } else if (tokens[0].equals("gush-crack")) {
                addObj(OBJ_GUSH_CRACK, x, NONE, false);
            } else if (tokens[0].equals("hydrant")) {
                addObj(OBJ_HYDRANT, x, NONE, false);
            } else if (tokens[0].equals("overhead-sign")) {
                addObj(OBJ_OVERHEAD_SIGN, x, token2, true);
            } else if (tokens[0].equals("rope")) {
                addObj(OBJ_ROPE_HORIZONTAL, x, NONE, false);
                addObj(OBJ_ROPE_VERTICAL, x, NONE, false);
            } else if (tokens[0].equals("spring")) {
                addObj(OBJ_SPRING, x, token2, true);
            } else if (tokens[0].equals("truck")) {
                addObj(OBJ_PARKED_TRUCK, x, NONE, false);
            } else if (tokens[0].equals("trigger-car-blue")) {
                addTrigger(x, CAR_BLUE);
            } else if (tokens[0].equals("trigger-car-silver")) {
                addTrigger(x, CAR_SILVER);
            } else if (tokens[0].equals("trigger-car-yellow")) {
                addTrigger(x, CAR_YELLOW);
            } else if (tokens[0].equals("trigger-hen")) {
                addTrigger(x, TRIGGER_HEN);
            } else if (tokens[0].equals("respawn-point")) {
                addRespawnPoint(x, token2);
            } else if (tokens[0].equals("deep-hole")) {
                addHole(HOLE_DEEP, x, token2);
            } else if (tokens[0].equals("passageway") ||
                        tokens[0].equals("passageway-arrow")) {

                addHole(HOLE_PASSAGEWAY_EXIT_CLOSED, x, token2);

                //Pushable crate over passageway entry
                addObj(OBJ_CRATE_PUSHABLE, x, NONE, false);
                ctx.pushableCrates[numPassageways - 1].obj = numObjs - 1;

                //Determine whether or not to show an arrow near the crate
                ctx.pushableCrates[numPassageways - 1].showArrow = false;
                if (tokens[0].equals("passageway-arrow")) {
                    ctx.pushableCrates[numPassageways - 1].showArrow = true;
                }
            } else {
                //Error: invalid object type
                return LVLERR_INVALID;
            }

            if (invalid) {
                return LVLERR_INVALID;
            }

            noObjects = false;
        }

        if (noObjects) {
            return LVLERR_INVALID;
        }

        //Error: running out of gushes due to gush cracks
        if (numGushes + numGushCracks > MAX_GUSHES) {
            return LVLERR_INVALID;
        }

        //Error: running out of positions in ctx.objs[] due to banana peels
        //thrown by triggered cars
        if (numObjs + numCarTriggers > MAX_OBJS) {
            return LVLERR_INVALID;
        }

        //Error: the number of respawn points is not the same as the number of
        //deep holes
        if (numRespawnPoints != numDeepHoles) {
            return LVLERR_INVALID;
        }

        //Ensure every respawn point is close enough to the corresponding deep
        //hole but not placed after it or over another deep hole
        for (i = 0, j = 0; i < numHoles; i++) {
            int hx = ctx.holes[i].x;
            int rx = ctx.respawnPoints[j].x;

            //Skip passageways
            if (ctx.holes[i].type != HOLE_DEEP) {
                continue;
            }

            if (rx >= hx || rx < hx - 4) {
                return LVLERR_INVALID;
            }

            if (i > 1) {
                int prevHole = i - 1;

                //Find previous deep hole (skip passageways)
                while (prevHole >= 0 && ctx.holes[prevHole].type != HOLE_DEEP) {
                    prevHole--;
                }

                if (prevHole >= 0) {
                    //Right side of previous deep hole
                    hx = ctx.holes[prevHole].x + ctx.holes[prevHole].width - 2;

                    if (rx <= hx) {
                        return LVLERR_INVALID;
                    }
                }
            }

            j++;
        }

        //First solid for the floor
        addSolid(SOL_FULL, 0, FLOOR_Y, ctx.levelSize, 80);

        //Convert positions from level blocks to pixels for deep holes and
        //passageways and adjust solids around them
        for (i = 0; i < numHoles; i++) {
            int prevObRight, obLeft, obWidth;
            boolean isDeep = (ctx.holes[i].type == HOLE_DEEP);

            x = ctx.holes[i].x;
            w = ctx.holes[i].width;

            prevObRight = x * LEVEL_BLOCK_SIZE;
            obLeft = (x + w) * LEVEL_BLOCK_SIZE;
            obWidth = ctx.levelSize - obLeft;

            if (isDeep) { //Deep hole
                prevObRight += 12;
                obLeft -= 8;
            } else { //Passageway
                prevObRight += 6;
            }

            //Adjust the previous floor solid so that it does not cover the
            //hole
            ctx.solids[i].right = prevObRight;

            //Add solid for the floor after the hole
            addSolid(SOL_FULL, obLeft, FLOOR_Y, obWidth, 80);

            //Too many solids
            if (invalid) {
                return LVLERR_INVALID;
            }

            //Convert position
            ctx.holes[i].x *= LEVEL_BLOCK_SIZE;
        }

        //Convert positions from level blocks to pixels and add solids for
        //objects in ctx.objs[]
        for (i = 0; i < numObjs; i++) {
            x = ctx.objs[i].x * LEVEL_BLOCK_SIZE;
            y = ctx.objs[i].y;
            if (y != NONE) {
                y *= LEVEL_BLOCK_SIZE;
            }

            switch (ctx.objs[i].type) {
                case OBJ_BANANA_PEEL:
                    x += 16;
                    y -= 8;
                    break;

                case OBJ_PARKED_CAR_BLUE:
                case OBJ_PARKED_CAR_SILVER:
                case OBJ_PARKED_CAR_YELLOW:
                    y = PARKED_CAR_Y;
                    addSolid(SOL_FULL, x + 4, y + 18, 20, 4);
                    addSolid(SOL_SLOPE_UP, x + 27, y + 2, 15, 0);
                    addSolid(SOL_VERTICAL, x + 48, y + 2, 16, 4);
                    addSolid(SOL_SLOPE_DOWN, x + 66, y + 2, 18, 0);
                    addSolid(SOL_KEEP_ON_TOP, x + 88, y + 20, 16, 4);
                    addSolid(SOL_KEEP_ON_TOP, x + 104, y + 22, 16, 4);
                    addSolid(SOL_FULL, x + 120, y + 24, 8, 4);
                    break;

                case OBJ_COIN_SILVER:
                case OBJ_COIN_GOLD:
                    x += 8;
                    break;

                case OBJ_CRATE_PUSHABLE:
                    y = PUSHABLE_CRATE_Y;
                    break;

                case OBJ_GUSH:
                    y = GUSH_INITIAL_Y;
                    break;

                case OBJ_GUSH_CRACK:
                    y = GUSH_CRACK_Y;
                    break;

                case OBJ_HYDRANT:
                    y = HYDRANT_Y;
                    addSolid(SOL_FULL, x + 4, y + 8, 8, 4);
                    break;

                case OBJ_OVERHEAD_SIGN:
                    y -= 8;
                    addSolid(SOL_FULL, x + 12, y, 4, 32);
                    break;

                case OBJ_ROPE_HORIZONTAL:
                    x += 10;
                    y = ROPE_Y;
                    break;

                case OBJ_ROPE_VERTICAL:
                    x += 32;
                    y = ROPE_Y + 5;
                    break;

                case OBJ_SPRING:
                    x += 8;
                    y += 8;
                    break;

                case OBJ_PARKED_TRUCK:
                    y = PARKED_TRUCK_Y;
                    addSolid(SOL_FULL, x, y + 4, 224, 96);
                    addSolid(SOL_FULL, x + 224, y + 23, 55, 80);
                    break;
            }

            //Too many solids
            if (invalid) {
                return LVLERR_INVALID;
            }

            //Apply converted position
            ctx.objs[i].x = x;
            ctx.objs[i].y = y;
        }

        //Set properties for ctx.pushableCrates[]
        for (i = 0; i < numPassageways; i++) {
            int obj = ctx.pushableCrates[i].obj;
            x = ctx.objs[obj].x;
            ctx.pushableCrates[i].x = x;
            ctx.pushableCrates[i].xmax = x + LEVEL_BLOCK_SIZE;
        }

        //Convert positions from level blocks to pixels and add solids for crate
        //blocks
        for (i = 0; i < numCrateBlocks; i++) {
            x = ctx.crateBlocks[i].x * LEVEL_BLOCK_SIZE;
            y = ctx.crateBlocks[i].y * LEVEL_BLOCK_SIZE;
            w = ctx.crateBlocks[i].width * LEVEL_BLOCK_SIZE;
            h = ctx.crateBlocks[i].height * LEVEL_BLOCK_SIZE;
            addSolid(SOL_FULL, x, y, w, h);

            //Too many solids
            if (invalid) {
                return LVLERR_INVALID;
            }

            //Apply converted position
            ctx.crateBlocks[i].x = x;
            ctx.crateBlocks[i].y = y;
        }

        //Convert respawn point positions from level blocks to pixels
        for (i = 0; i < numRespawnPoints; i++) {
            ctx.respawnPoints[i].x *= LEVEL_BLOCK_SIZE;
            ctx.respawnPoints[i].x += 3;

            ctx.respawnPoints[i].y *= LEVEL_BLOCK_SIZE;
            ctx.respawnPoints[i].y -= 12;
        }

        //Convert trigger positions from level blocks to pixels
        for (i = 0; i < numTriggers; i++) {
            ctx.triggers[i].x *= LEVEL_BLOCK_SIZE;
        }

        //Add solids for passageways and pushable crates over passageway entries
        //
        //There is exactly one pushable crate for each passageway
        for (i = 0, j = 0; i < numHoles; i++) {
            int sol;

            if (ctx.holes[i].type == HOLE_DEEP) {
                continue;
            }

            x = ctx.holes[i].x;
            w = ctx.holes[i].width * LEVEL_BLOCK_SIZE;

            //Bottom solid
            addSolid(SOL_FULL, x, PASSAGEWAY_BOTTOM_Y, w, 4);

            //Top solid
            addSolid(SOL_FULL, x + LEVEL_BLOCK_SIZE, FLOOR_Y, w - 46, 13);

            //Entry and exit solids, which prevent the player character from
            //leaving the passageway through the entry or entering it through
            //the exit
            addSolid(SOL_PASSAGEWAY_ENTRY, x + 6, FLOOR_Y, 18, 13);
            addSolid(SOL_PASSAGEWAY_EXIT, x + w - 22, FLOOR_Y, 22, 13);

            ctx.holes[i].x = x;

            //Pushable crate solid
            x = (int)ctx.pushableCrates[j].x;
            y = PUSHABLE_CRATE_Y;
            sol = addSolid(SOL_FULL, x, y, CRATE_WIDTH, CRATE_HEIGHT);
            ctx.pushableCrates[j].solid = sol;
            j++;

            //Too many solids
            if (invalid) {
                return LVLERR_INVALID;
            }
        }

        return LVLERR_NONE;
    }

    //--------------------------------------------------------------------------

    void addObj(int type, int x, int y, boolean useY) {
        int i;

        //Check if there are too many objects
        if (numObjs >= MAX_OBJS) {
            invalid = true;
            return;
        }

        //Check if the object's position is within the allowed range
        if (y > 15 || (y != NONE && y < 3) || (useY && y == NONE) || x > xMax) {
            invalid = true;
            return;
        }

        //Check object repetition
        for (i = 0; i < numObjs; i++) {
            Obj obj = ctx.objs[i];

            if (obj.type == type && obj.x == x && obj.y == y) {
                invalid = true;
                return;
            }
        }

        ctx.objs[numObjs].type = type;
        ctx.objs[numObjs].x = x;
        ctx.objs[numObjs].y = y;

        numObjs++;
    }

    void addCrateBlock(int x, int y, int w, int h) {
        int x2 = x + w - 1;
        int y2 = y + h - 1;
        int i;

        //Check if there are too many crate blocks
        if (numCrateBlocks >= MAX_CRATE_BLOCKS) {
            invalid = true;
            return;
        }

        //Check if the crate block's position is within the allowed range
        if (x > xMax - 2 || y < 3 || y > 15) {
            invalid = true;
            return;
        }

        //Check if the crate block's size is within the allowed range
        if (w < 1 || w > 4 || h < 1 || h > 5) {
            invalid = true;
            return;
        }

        if (x2 > xMax - 2) {
            //Error: crate block width extends beyond or too close to level's
            //right boundary
            invalid = true;
            return;
        }

        //Check crate block repetition or overlap
        for (i = 0; i < numCrateBlocks; i++) {
            CrateBlock ct = ctx.crateBlocks[i];
            int cx2 = ct.x + ct.width - 1;
            int cy2 = ct.y + ct.height - 1;

            if (ct.x == x && ct.y == y) {
                invalid = true;
                return;
            }

            if (x <= cx2 && y2 >= ct.y && y <= cy2) {
                invalid = true;
                return;
            }
        }

        ctx.crateBlocks[numCrateBlocks].x = x;
        ctx.crateBlocks[numCrateBlocks].y = y;
        ctx.crateBlocks[numCrateBlocks].width = w;
        ctx.crateBlocks[numCrateBlocks].height = h;

        numCrateBlocks++;
    }

    //Adds a deep hole or a passageway
    void addHole(int type, int x, int w) {
        int x2 = x + w - 1;
        int maxWidth = (type == HOLE_DEEP) ? 16 : 32;
        int i;

        //Check if there are too many holes
        if (numHoles >= MAX_HOLES) {
            invalid = true;
            return;
        }

        //Check if the hole's position and size are within the allowed range
        if (x > xMax - 2 || w < 2 || w > maxWidth) {
            invalid = true;
            return;
        }

        if (x2 > xMax - 2) {
            //Error: hole width extends beyond or too close to level's right
            //boundary
            invalid = true;
            return;
        }

        //Check hole repetition or overlap
        for (i = 0; i < numHoles; i++) {
            int hx1 = ctx.holes[i].x;
            int hx2 = hx1 + ctx.holes[i].width - 1;

            if (hx1 == x || x <= hx2) {
                invalid = true;
                return;
            }
        }

        ctx.holes[numHoles].type = type;
        ctx.holes[numHoles].x = x;
        ctx.holes[numHoles].width = w;

        if (type == HOLE_DEEP) {
            numDeepHoles++;
        } else {
            numPassageways++;

            //Too many passageways
            if (numPassageways > MAX_PASSAGEWAYS) {
                invalid = true;
                return;
            }
        }

        numHoles++;
    }

    void addRespawnPoint(int x, int y) {
        int i;

        //Check if there are too many triggers
        if (numRespawnPoints >= MAX_RESPAWN_POINTS) {
            invalid = true;
            return;
        }

        //Check if the respawn point's position is within the allowed range
        if (x > xMax || y < 3 || y > 15) {
            invalid = true;
            return;
        }

        //An X position cannot be shared by two or more respawn points
        for (i = 0; i < numRespawnPoints; i++) {
            RespawnPoint rp = ctx.respawnPoints[i];

            if (rp.x == x) {
                invalid = true;
                return;
            }
        }

        ctx.respawnPoints[numRespawnPoints].x = x;
        ctx.respawnPoints[numRespawnPoints].y = y;

        numRespawnPoints++;
    }

    void addTrigger(int x, int what) {
        int i;

        //Check if there are too many triggers
        if (numTriggers >= MAX_TRIGGERS) {
            invalid = true;
            return;
        }

        //Check if the trigger's position is within the allowed range
        if (x > xMax - 20) {
            invalid = true;
            return;
        }

        //Check trigger repetition or excessive proximity
        for (i = 0; i < numTriggers; i++) {
            int tx = ctx.triggers[i].x;

            if (tx == x || tx > x - 28) {
                invalid = true;
                return;
            }
        }

        ctx.triggers[numTriggers].x = x;
        ctx.triggers[numTriggers].what = what;

        numTriggers++;

        if (what != TRIGGER_HEN) {
            numCarTriggers++;
        }
    }

    int addSolid(int type, int x, int y, int width, int height) {
        //Check if there are too many solids
        if (numSolids >= MAX_SOLIDS) {
            invalid = true;
            return -1;
        }

        //Slopes are always contained in a square, which means that the height
        //equals the width
        if (type == SOL_SLOPE_UP || type == SOL_SLOPE_DOWN) {
            height = width;
        }

        ctx.solids[numSolids].type = type;
        ctx.solids[numSolids].left = x;
        ctx.solids[numSolids].right = x + width;
        ctx.solids[numSolids].top = y;
        ctx.solids[numSolids].bottom = y + height;
        numSolids++;

        return numSolids - 1;
    }
}

